enhancements/machine-config/install-time-support-for-image-mode-on-openshift.md

```diff
@@ -0,0 +1,342 @@
+2. Bootstrap MCO/MCC parses the ignition file to generate the necessary MachineConfigs
+3. Bootstrap MCO/MCC pods establishes connection with external image registry (for this enhancement, running a registry on the bootstrap node is a non-goal)
+4. Once the rendered-MC is generated, bootstrap MCO/MCC starts the "on-cluster" image build process by creating a pod to build the custom OS image for the master MCP and push the built image to the external registry
+5. Bootstrap MCO/MCC creates a MachineConfig to override the `osImageURL` to point to the newly built image for the master pool
+6. Bootstrap node creates the `control plane` nodes (master pool) with the stock AMI
+7. Bootstrap MCO runs `bootc switch` or `rpm-ostree rebase` to boot the master nodes into the custom built OS image refered to by the `osImageURL` value
+
+#### Step 3: Worker Node Deployment
+
+1. Wait for the rendered-MC for the worker pool to be generated
+2. Wait for all the configs defined during install time to be applied to the cluster - the worker MOSC should be part of this
+3. Control plane nodes parse the worker MOSC configuration to start the "on-cluster" build process similar to how it is currently done for post-install image mode
+4. The control plane node which is running the MachineOSBuilder pod establishes connection with external image registry
+5. The control plane node which is running the MachineOSBuilder pod pushes the built image to the registry
+6. MCO/MCC creates a MachineConfig to override the `osImageURL` for the worker pool to point to the newly built image
+7. The worker nodes are created with the stock AMI
+8. The MCO runs `bootstrap switch` or `rpm-ostree rebase` to boot the worker nodes into the custom built OS image
+
+#### Notes and Restrictions:
+
+1. The Bootstrap node cannot communicate with the worker nodes
+2. The MOSC and MOSB for the master pool will have to be mocked and applied to the cluster so that when the MachineOSBuilder is started, it knows the history of the image built and applied to the master pool so we don't end up rebuilding unnecessarily
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+The primary use-case for Image Mode in Hypershift is for the host cluster admins to quickly roll out hotfixes and other changes to all of the guest cluster nodes (in addition to control plane nodes). The guest cluster owners do not need to customize their OS. Some adaptations would be required to Hypershift in order to traverse the host <-> guest cluster barrier. It is worth noting that within a Hypershift setup, the management cluster may be several versions behind the guest clusters or may not even be running OpenShift at all. Additionally, the guest clusters can each run different versions of OCP, meaning that it would not be appropriate to build a single base image to be used for all guest cluster nodes. Instead, images should be built for each guest clusters' OS / extensions image.
```

Comment:
I think we should remove this paragraph since it doesn't directly pertain to the install-time configuration of HCP.

Reply:
Fixed